File: templates/zig/runner.zig
//! Gota runner (Zig 0.16) - YOUR code. This is the file you edit.
//!
//! It declares the operation(s) to measure and plugs them into the harness. There is
//! no timing, batching, or JSON here: `register` receives a `*gota.Bencher` and calls
//! `b.bench(name, op)` once per operation. Each op is `fn([]u8) void` over the buffer.
//!
//!     zig build-exe runner.zig -O ReleaseFast
//!     ./runner [buffer_bytes] [warmup_seconds] [measure_seconds]

const std = @import("std");
const gota = @import("gota.zig");

const IMPL = "example"; // your implementation's name

// Replace with your real operation. It receives the shared buffer. This example writes
// the buffer in place, which is always observed. If your op instead computes a value,
// consume it (e.g. `data[0] ^= @truncate(result)`) or ReleaseFast may delete the work.
fn exampleOp(data: []u8) void {
    for (data) |*x| x.* ^= 0x5A;
}

fn register(b: *gota.Bencher) !void {
    try b.bench("example", exampleOp);
}

pub fn main(init: std.process.Init) !void {
    try gota.run(init, IMPL, register);
}

File: templates/ts/runner.ts
// Gota runner (TypeScript) - YOUR code. This is the file you edit.
//
// It declares the operation(s) to measure and plugs them into the harness. There is
// no timing, batching, or JSON here: run hands your callback a Bencher and the
// buffer, and you call b.bench(name, op) once per operation.
//
//   npx tsx runner.ts [buffer_bytes] [warmup_seconds] [measure_seconds]
//   # or, with no extra tooling on Node >= 22.6:
//   node --experimental-strip-types runner.ts [buffer_bytes] [warmup_seconds] [measure_seconds]

import { run, type Bencher } from "./gota.ts";

const IMPL = "example"; // your implementation's name

run(IMPL, (b: Bencher, data: Uint8Array) => {
  // Replace with your real operation(s). Each op closes over `data` and runs the
  // work to measure. Call b.bench once per operation you want in the table.
  //
  // This example writes `data` in place, which is always observed. If your op instead
  // computes a value, consume it (write a byte back into `data`) so the JIT can't elide
  // the work and you measure nothing.
  b.bench("example", () => {
    for (let i = 0; i < data.length; i++) data[i] ^= 0x5a;
  });
});
